frontend/src/services/authentication_handler_service/authentication_handler_service.ts

```diff
@@ -12,34 +11,67 @@ export class FailureToSignInError extends Error {
 }
 
 /**
- * Communicates with gapi to retrieve auth code from user.
- * Exchanges auth code and redirect URI for ID token from backend.
+ * Communicates with gapi to manage a user's login status
+ * and retrieve auth codes / tokens.
  */
 export class AuthenticationHandlerService {
+  token: string | undefined = undefined;
+  authInstance: gapi.auth2.GoogleAuth | undefined = undefined;
+
   /** Backend is responsible for retrieving the ID token. */
-  constructor(private readonly backend: BackendAuthenticationInterface) {
-    // Safe because the gapi script is loaded in the head with no async/defer.
-    // TODO: Make gapi script async and call this when it has loaded
-    window.gapi.load('auth2', () => {
-      window.gapi.auth2.init({
-        client_id: CLIENT_ID,
-      });
-    });
+  constructor(private readonly backend: BackendAuthenticationInterface) {}
+
+  /**
+   * Determines if a user is logged in.
+   * @return true if user is logged in, false otherwise
+   */
+  getUserLoginStatus(): boolean {
+    const auth = this.getAuthInstance();
+    return auth.isSignedIn.get() && this.getToken() !== undefined;
   }
 
   /**
-   * Signs the user in, retrieving a token from the backend.
+   * Gets the parsed ID token of the user.
+   * @return the parsed token if it exists, otherwise the empty object
+   */
+  getParsedToken() {
```

Comment:
Would be awesome for this to return a specified type, like `TokenInterface` or something

Reply:
TokenInterface has been created! Very good advice.